pyiceberg/table/__init__.py

```diff
@@ -277,6 +279,50 @@ def __init__(self, table: Table, autocommit: bool = False):
         self._autocommit = autocommit
         self._updates = ()
         self._requirements = ()
+        self._manage_snapshots = Transaction.ManageSnapshot(self)
+
+    class ManageSnapshot:
```

Comment:
> I am thinking that ManageSnapshots can be an outer class like [UpdaetSchema](https://github.com/apache/iceberg-python/blob/55bbc1863aba6646b2dc36573355aabb11909095/pyiceberg/table/__init__.py#L1920) and implements the UpdateTableMetadata:

I agree, that's a great catch. I've now updated the class. 

> We could utilize Transaction's autocommit option when implementing table.manage_snapshots, like:

Updated!

> We could also make create_tag returns ManageSnapshots itself to support chain operations:

This is great, updated to allow method chaining!

> There are several benefits:

I do agree with the benefits. However, with this change, `ManageSnapshots` would be specific to metadata update operations only. In my other PRs, I've implemented `cherrypick` and `publish_changes`, both of which need to create a new snapshot from an existing snapshot and using `_MergingSnapshotProducer`. I'm assuming we would like to have these APIs under `ManageSnapshots` like `table.manage_snapshots().cherrypick(snapshot_id)`. 

Ex:
```
class ManageSnapshots:
  def cherrypick(self, snapshot_id: int) -> ManageSnapshots:
      CherryPickSnapshot(snapshot_id, .....).cherrypick().commit() # we can also enable autocommit here
      return self
```

Here, we are performing cherrypick through a class CherryPickSnapshot that implements _MergingSnapshotProducer. So no updates or requirements are added to ManageSnapshots, so ManageSnapshots.commit() is never used. Would this be a reasonable implementation?

Reply:
Thanks for the update! 

> So no updates or requirements are added to ManageSnapshots and ManageSnapshots.commit() doesn't update the table

Yeah, I think that will still be a reasonable implementation. Just one thing to note: in `cherrypick` operation we need to commit any pending snapshot ref updates before cherrypick. For example:
```python
class ManageSnapshots:
  def cherrypick(self, snapshot_id: int) -> ManageSnapshots:
      self.commit()
      CherryPickSnapshot(snapshot_id, .....).cherrypick()
      return self
```
We could also add other helper `UpdateTableMetadata` implementation if needed and make them private.

Reply:
I'll keep note of this for the cherrypick and other similar PRs!

Reply:
Thanks for splitting this into a separate class. I think that's a great idea and also reflects the [UpdateSnapshotReferencesOperation](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java) in Java. Here they keep state inside of the transaction, so there you can do:

```python
with table.manage_snapshots() as ms:
       ms.create_tag("production-3-6-2024", table.current_snapshot().snapshot_id)
       ms.drop_tag("production-3-6-2024")
```

This would then result in a no-op. We don't have to do this for this first iteration.

> We could do multiple tags and branch creations and then _apply once to reduce the number of transaction._apply() call since that will trigger requirement validations every time. This also allow us to avoid duplicated requirements being added such as AssertTableUUID

I'm unsure where the `AssertTableUUID` comes from, since I don't think that's required when updating a ref. Looks like this isn't checked in Java: https://github.com/apache/iceberg/blob/252168419f8ac1d251b39f0944a189184056e543/core/src/main/java/org/apache/iceberg/UpdateRequirements.java#L113-L125

Reply:
Oh I added it after a review in this PR. [Review comment](https://github.com/apache/iceberg-python/pull/728#discussion_r1597761797).

Reply:
Thanks, I see. Let's iterate with the community to understand why we need that requirement. 👍